executor/infoschema_reader.go

```diff
@@ -1306,6 +1312,177 @@ func (e *memtableRetriever) setDataFromTableConstraints(ctx sessionctx.Context,
+		return rows, nil
+	}
+	retRows := make([][]types.Datum, len(rows))
+	for i, fullRow := range rows {
+		row := make([]types.Datum, len(e.outputCols))
+		for j, col := range e.outputCols {
+			row[j] = fullRow[col.Offset]
+		}
+		retRows[i] = row
+	}
+	return retRows, nil
+}
+
+type initialTable struct {
+	db string
+	tb *model.TableInfo
+}
+
+func (e *tableStorageStatsRetriever) initialize(sctx sessionctx.Context) error {
+	is := infoschema.GetInfoSchema(sctx)
+	dbs := is.AllSchemas()
+	sort.Sort(infoschema.SchemasSorter(dbs))
+	schemas := e.extractor.TableSchema
+	tables := e.extractor.TableName
+	if len(schemas) != 0 {
+		for _, schema := range dbs {
+			if schemas.Exist(schema.Name.L) {
+				if len(tables) != 0 {
+					for _, table := range schema.Tables {
+						if tables.Exist(table.Name.L) {
+							e.initialTables = append(e.initialTables, &initialTable{db: schema.Name.L, tb: table})
+						}
+					}
+				} else {
+					for _, table := range schema.Tables {
+						e.initialTables = append(e.initialTables, &initialTable{db: schema.Name.L, tb: table})
+					}
+				}
+			}
+
+		}
+		if len(e.initialTables) == 0 && tables == nil {
+			return errors.Errorf("schema or table not exist, please check the schema and table")
+		}
+	} else {
+		if len(tables) != 0 {
+			for _, schema := range dbs {
+				for _, table := range schema.Tables {
+					if tables.Exist(table.Name.L) {
+						e.initialTables = append(e.initialTables, &initialTable{db: schema.Name.L, tb: table})
+					}
+				}
+			}
+		} else {
+			for _, schema := range dbs {
+				for _, table := range schema.Tables {
+					e.initialTables = append(e.initialTables, &initialTable{schema.Name.L, table})
+				}
+			}
+		}
+	}
+
+	// Cache the PD address.
+	tikvStore, ok := sctx.GetStore().(tikv.Storage)
+	if !ok {
+		return errors.New("Information about TiKV region status can be gotten only when the storage is TiKV")
+	}
+	etcd, ok := tikvStore.(tikv.EtcdBackend)
+	if !ok {
+		return errors.New("not implemented")
+	}
+	pdAddrs := etcd.EtcdAddrs()
+	if len(pdAddrs) == 0 {
+		return errors.New("pd unavailable")
+	}
+	e.pdAddress = pdAddrs[0]
+	e.initialized = true
+	return nil
+}
+
+// PdRegionStats is the json response from PD.
+type PdRegionStats struct {
+	Count          int            `json:"count"`
+	EmptyCount     int            `json:"empty_count"`
+	StorageSize    int64          `json:"storage_size"`
+	StorageKeys    int64          `json:"storage_keys"`
+	StorePeerCount map[uint64]int `json:"store_peer_count"`
+}
+
+func (e *tableStorageStatsRetriever) setDataForTableStorageStats(ctx sessionctx.Context) ([][]types.Datum, error) {
+	rows := make([][]types.Datum, 0, 1024)
+
+	count := 0
+	for i := e.curTable; e.curTable < len(e.initialTables) && count < 1024; i++ {
+		table := (e.initialTables)[e.curTable]
+		tableID := table.tb.ID
+		// Include table and index data, because their range located in tableID_i tableID_r
+		startKey := tablecodec.EncodeTablePrefix(tableID)
```

Comment:
For the system table in `INFORMATION_SCHEMA`, `METRICS_SCHEMA`, `PERFORMANCE_SCHEMA`, the `peer count`, `region count`, `table size` and so on should be `NULL`?

Reply:
Why null? It seems these sys table would share the region and return the same value of region's size and keys.